datafusion/macros/src/metric_doc.rs

```diff
@@ -0,0 +1,243 @@
+    }
+    .into())
+}
+
+fn documented_exec_impl(input: TokenStream) -> syn::Result<TokenStream> {
+    let input = syn::parse::<DeriveInput>(input)?;
+    let ident = input.ident;
+    let doc = doc_expr(&input.attrs);
+    let metrics = metrics_list(&input.attrs)?;
+
+    let metrics_refs = metrics.iter().map(|metric| {
+        quote! { <#metric as datafusion_doc::metric_doc_sections::DocumentedMetrics>::DOC }
+    });
+
+    let mod_ident = format_ident!("__datafusion_doc_exec_{}", ident);
+
+    Ok(quote! {
+        #[allow(non_snake_case)]
+        mod #mod_ident {
+            use super::*;
+
+            static DOCUMENTATION: datafusion_doc::metric_doc_sections::ExecDoc =
+                datafusion_doc::metric_doc_sections::ExecDoc {
+                    name: stringify!(#ident),
+                    doc: #doc,
+                    metrics: &[#(#metrics_refs),*],
+                };
+
+            impl datafusion_doc::metric_doc_sections::DocumentedExec for super::#ident {
+                fn exec_doc(
+                ) -> &'static datafusion_doc::metric_doc_sections::ExecDoc {
+                    &DOCUMENTATION
+                }
+            }
+
+            datafusion_doc::metric_doc_sections::inventory::submit! {
+                datafusion_doc::metric_doc_sections::ExecDocEntry(&DOCUMENTATION)
+            }
+        }
+    }
+    .into())
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum MetricDocPosition {
+    Common,
+    Operator,
+}
+
+fn metrics_position(attrs: &[Attribute]) -> syn::Result<MetricDocPosition> {
+    let mut position = MetricDocPosition::Operator;
+    for attr in attrs {
+        if is_metrics_attr(attr) {
+            attr.parse_nested_meta(|meta| {
+                if meta.path.is_ident("common") {
+                    position = MetricDocPosition::Common;
+                    return Ok(());
+                }
+                Err(meta.error("unsupported metric_doc attribute"))
+            })?;
+        }
+    }
+    Ok(position)
+}
+
+fn metrics_list(attrs: &[Attribute]) -> syn::Result<Vec<Path>> {
+    let mut metrics = vec![];
+    for attr in attrs {
+        if is_metrics_attr(attr) {
+            attr.parse_nested_meta(|meta| {
+                metrics.push(meta.path);
+                Ok(())
+            })?;
+        }
+    }
+
+    Ok(metrics)
+}
+
+fn doc_expr(attrs: &[Attribute]) -> proc_macro2::TokenStream {
+    let mut doc_lines = Vec::new();
+
+    for attr in attrs {
+        if !attr.path().is_ident("doc") {
+            continue;
+        }
+
+        if let syn::Meta::NameValue(meta) = &attr.meta
+            && let Expr::Lit(ExprLit {
+                lit: Lit::Str(lit_str),
+                ..
+            }) = &meta.value
+        {
+            doc_lines.push(lit_str.value());
+        }
+    }
+
+    if doc_lines.is_empty() {
+        quote!("")
+    } else {
+        let sanitized = doc_lines
+            .into_iter()
+            .map(|line| line.trim().to_string())
```

Comment:
```suggestion
            .map(|line| line.trim_end().to_string())
```
Preserve any whitespace used intentionally for indentation.